File: src/main/java/no/nav/samordning/hendelser/VaultHikariConfig.java
package no.nav.samordning.hendelser;

import com.zaxxer.hikari.HikariDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import org.springframework.vault.core.lease.LeaseEndpoints;
import org.springframework.vault.core.lease.SecretLeaseContainer;
import org.springframework.vault.core.lease.domain.RequestedSecret;
import org.springframework.vault.core.lease.event.SecretLeaseCreatedEvent;

@Configuration
@Profile("!test")
public class VaultHikariConfig implements InitializingBean {

    private SecretLeaseContainer container;
    private HikariDataSource hikariDataSource;

    @Value("${VAULT_POSTGRES_BACKEND}")
    private String vaultPostgresBackend;
    @Value("${VAULT_POSTGRES_ROLE}")
    private String vaultPostgresRole;

    private static Logger LOGGER = LoggerFactory.getLogger(VaultHikariConfig.class.getName());

    public VaultHikariConfig(SecretLeaseContainer container, HikariDataSource hikariDataSource) {
        this.container = container;
        this.hikariDataSource = hikariDataSource;
    }

    @Override
    public void afterPropertiesSet() {
        container.setLeaseEndpoints(LeaseEndpoints.SysLeases); // TODO: Remove
        RequestedSecret secret = RequestedSecret.rotating(this.vaultPostgresBackend + "/creds/" + this.vaultPostgresRole);
        container.addLeaseListener(leaseEvent -> {
            if (leaseEvent.getSource() == secret && leaseEvent instanceof SecretLeaseCreatedEvent) {
                LOGGER.info("Rotating creds for path: " + leaseEvent.getSource().getPath());
                SecretLeaseCreatedEvent slce = (SecretLeaseCreatedEvent) leaseEvent;
                String username = slce.getSecrets().get("username").toString();
                String password = slce.getSecrets().get("password").toString();
                hikariDataSource.setUsername(username);
                hikariDataSource.setPassword(password);
                hikariDataSource.getHikariConfigMXBean().setUsername(username);
                hikariDataSource.getHikariConfigMXBean().setPassword(password);
            }
        });
        container.addRequestedSecret(secret);
    }
}

File: src/main/kotlin/no/nav/samordning/hendelser/VaultHikariConfig.kt
package no.nav.samordning.hendelser

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import no.nav.vault.jdbc.hikaricp.HikariCPVaultUtil
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile


@Configuration
@Profile("!test")
class VaultHikariConfig(
    @Value("\${SPRING_DATASOURCE_URL}")
    private val vaultPostgresUrl: String,
    @Value("\${VAULT_POSTGRES_BACKEND}")
    private val vaultPostgresBackend: String,
    @Value("\${VAULT_POSTGRES_ROLE}")
    private val vaultPostgresRole: String,
    @Value("\${MAX_POOL_SIZE}")
    private val maxPoolSize: Int
) {

    private val hikariConfig = HikariConfig().apply {
        jdbcUrl = vaultPostgresUrl
        minimumIdle = 0
        maxLifetime = 30001
        maximumPoolSize = maxPoolSize
        connectionTimeout = 250
        idleTimeout = 10001
    }

    @Bean
    fun hikariDataSource(): HikariDataSource = HikariCPVaultUtil.createHikariDataSourceWithVaultIntegration(
        hikariConfig,
        vaultPostgresBackend,
        vaultPostgresRole
    )
}
